Update descendant inherited fields refs #11855

MikeFE opened the pull request:
For fields such as creator, repository and 'part of', descendant records were
not being updated on-the-fly to reflect said inherited fields from their parent
records when they were updated. This fix adds an async job to update descendant
records when the parent changes the mentioned fields.

Commit:
Update descendant inherited fields refs #11855

Commit:
Break out Propel connection singleton

We can remove the connection function call from populate as well since it uses
QubitPdo now which will ensure we have a SQL connection instance.

jraddaoui commented on lib/job/arUpdateDescendantsJob.class.php:
```diff
@@ -0,0 +1,54 @@
+ */
+
+/**
+ *
```
Extra line.

jraddaoui commented on lib/job/arUpdateDescendantsJob.class.php:
```diff
@@ -0,0 +1,54 @@
+    $searchIo = new arElasticSearchInformationObject;
+    $searchIo->recursivelyUpdateInformationObjects($parameters['tldId'], count($this->topLevelDesc->descendants),
+      'updateInheritedFields');
+
```
Extra line.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,56 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    if ($data)
+    {
+      $data['_id'] = $item->id;
+      QubitSearch::getInstance()->addDocument($data, 'QubitInformationObject');
```
This will replace the existing document with a new one that will only have the inherited data, everything else will be removed. Please, take a look at `partialUpdate`:

https://github.com/artefactual/atom/blob/qa/2.5.x/plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php#L456-L486

As noted in that function comment, the way the merge works in `updateDocument` means that, if a field is not present in the new data, it won't be deleted from the document, so it should be set to `null` to do so.

Right now, only the transcript fields are expecting partial updates. See the following comment:

https://github.com/artefactual/atom/blob/qa/2.5.x/plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php#L1070-L1075

And how the partial update is done from the finding aid job:

https://github.com/artefactual/atom/blob/qa/2.5.x/lib/job/arFindingAidJob.class.php#L186
https://github.com/artefactual/atom/blob/qa/2.5.x/lib/job/arFindingAidJob.class.php#L275
https://github.com/artefactual/atom/blob/qa/2.5.x/lib/job/arFindingAidJob.class.php#L319-L326

If we handle `null` values this way, most likely we won't need changes in the queries and aggregations made over those fields, but we should double check. We may need to use empty arrays instead as this are "multifields".

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php:
```diff
@@ -1334,10 +1377,10 @@ public function serialize()
+      if (null !== $partOf = $this->getPartOf($collectionRootId))
+      {
+        $serialized['partOf'] = $partOf;
+      }
```
We are doing things twice in here. You could replace the entire `if` wrapping this one with this one and remove the `$collectionRootId = $this->getCollectionRootId();` call made before the `if`, as you're getting it in the new `getPartOf` function, which also makes unnecessary to pass it to the function.

MikeFE replied:
ah good call, yeah I missed that when refactoring

jraddaoui commented on apps/qubit/modules/informationobject/actions/editAction.class.php:
```diff
@@ -597,6 +597,8 @@ protected function processForm()
       $this->getUser()->setFlash('notice', $message);
     }
 
```
Like with the publication status change made on top and like we do with other jobs, could you add a flash notice telling that job has been started to update the descendants?

MikeFE replied:
agreed, I will implement it

jraddaoui commented on apps/qubit/modules/informationobject/actions/editAction.class.php:
```diff
@@ -597,6 +597,8 @@ protected function processForm()
       $this->getUser()->setFlash('notice', $message);
     }
 
+    QubitJob::runJob('arUpdateDescendantsJob', array('tldId' => $this->resource->id));
```
What about other places where a description could be edited? Shouldn't we call the job from the IO save method?

We should also avoid the job execution if the description has no descendants, checking that `rgt - lft > 1`, like in:

https://github.com/artefactual/atom/blob/qa/2.5.x/apps/qubit/modules/informationobject/templates/updatePublicationStatusSuccess.php#L29

jraddaoui commented on lib/job/arUpdateDescendantsJob.class.php:
```diff
@@ -0,0 +1,54 @@
+ * @subpackage jobs
+ */
+
+class arUpdateDescendantsJob extends arBaseJob
```
I think we could improve a lot this job to accept other big ES documents updates without a big amount of changes.

I'd do the following:
- Rename the job to arUpdateIoEsDocumentsJob (or similar).
- Require the following parameters:
  - An array of IO ids.
  - A boolean to update the IOs from the array or not.
  - A boolean to update their descendants or not.
- Change the process logic to:
  - Update the IOs from the array based on the first boolean.
  - Update their descendants based on the second boolean.

That way we could use this job in this case and in other big IO updates, like when a term or an actor is updated:

https://github.com/artefactual/atom/blob/qa/2.5.x/apps/qubit/modules/term/actions/editAction.class.php#L578-L589
https://github.com/artefactual/atom/blob/qa/2.5.x/lib/model/QubitActor.php#L146-L153

jraddaoui replied:
Ideally, this job could accept a partial update option, where only some fields are updated instead of the entire IOs, but that could be done in the future. This option should only be considered in the IOs from the array, descendants should always use partial updates, like it's being done.

Later, we could rename and add other resources to this job, like QubitActor and QubitTerm.

jraddaoui replied:
Feel free to disregard this if running out of time. We can create a new feature ticket to not forgot about the enhancement in the future.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php:
```diff
@@ -1063,6 +1063,49 @@ private function getActRights()
+      $rootSlug = QubitPdo::fetchColumn('SELECT slug FROM slug WHERE object_id=?', array($collectionRootId));
+      if (!$rootSlug)
+      {
+        throw new sfException("No slug found for information object $collectionRootId");
```
Throwing an exception in here will avoid the entire document to be added to the ES index. I think it's better to have it without part of the `parOf` data than not having the document in the index.

MikeFE replied:
The first condition in the if is from a result from a db query indicating an info object exists. If one exists without a slug and isn't ROOT_ID, I think that's data corruption? I don't think it's wise to continue the code path here.

jraddaoui replied:
I agree this is a data corruption issue on the ancestor, but I don't think it's enough to remove all the descendants from the index. The search populate task will show an error at the end about the ancestor issue but, if this is called from the GUI, and somehow the ancestor lost its slug, you may end up loosing a lot of records without noticing.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php:
```diff
@@ -1063,6 +1063,49 @@ private function getActRights()
+                                                array($collectionRootId));
+      if (!$rootSourceCulture)
+      {
+        throw new sfException("No source culture found for information object $collectionRootId");
```
Same here.

jraddaoui reviewed (changes requested):
Hi @MikeFE,

We could do a big improvement to the job that should not require too much time and it will allow us to improve other parts of the application where a lot of ES documents are being updated.

See also my other comments bellow.

Commit:
Remove extra lines

Commit:
Remove redundant getPartOf parameter

Commit:
Set flash notification for update descendants job

Commit:
Do not start update descendants job if none

Commit:
Make updates use partial update functionality

Commit:
Avoid throwing exception over field missing

Source culture isn't very critical so throwing an exception here was undesired.

jraddaoui commented on apps/qubit/modules/informationobject/actions/editAction.class.php:
```diff
@@ -816,4 +817,26 @@ private function incrementMaskCounter()
+
+    $message = $this->context->i18n->__('Updating inherited fields for descendants. Check %1%job %2%%3% to view the status of the job.', $params);
+    $this->getUser()->setFlash('notice', $message);
+  }
```
I still think the job should be triggered from the IO save, as we now have an 'update on match' type of import and other places that may be making changes to the IOs.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+  }
+
+  /**
+   * Update ES fields descendants inherit from their ancestors. This assumes QubitInformationObject atm.
```
We are in arElasticSearchInformationObject ;)

jraddaoui commented on lib/job/arUpdateDescendantsJob.class.php:
```diff
@@ -0,0 +1,53 @@
+    $searchIo = new arElasticSearchInformationObject;
+    $searchIo->recursivelyUpdateInformationObjects($parameters['tldId'], count($this->topLevelDesc->descendants),
+      'updateInheritedFields', array('repository' => $this->topLevelDesc->getRepository(),
+      'inheritedCreators' => $this->topLevelDesc->getCreators()));
```
Cosmetic change: at first glace it makes me think `updateInheritedFields` and `inheritedCreators` belongs to the same array, I'd put the array in a new line.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    $node = new arElasticSearchInformationObjectPdo($item->id, $options);
+    $data = array();
+
+    if (isset($options['repository']) && null !== $repository = $node->getRepository())
```
Do you really need to check it's in the options array?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    if (isset($options['repository']) && null !== $repository = $node->getRepository())
+    {
+      $data['repository'] = arElasticSearchRepository::serialize($repository);
+    }
```
Otherwise set `$data['repository']` to `null` or empty array (you'll need to check what's better in this 'multifields')

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+      $data['repository'] = arElasticSearchRepository::serialize($repository);
+    }
+
+    if (isset($options['inheritedCreators']))
```
Again, not sure why you need to check this.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+      {
+        if (!isset($data['inheritedCreators']))
+        {
+          $data['inheritedCreators'] = array();
```
You'll need an empty array or null in the end, so I'd create it before the foreach to avoid this if.

jraddaoui replied:
Also it's called `creators` in the index, not `inheritedCreators`. This should have failed while testing the changes.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+
+    if (isset($options['inheritedCreators']))
+    {
+      foreach ($options['inheritedCreators'] as $creator)
```
You're going only over the parent creators in this foreach. In the index and in the GUI we only inherit the parent creators if the IO doesn't have creators by itself. Check:

https://github.com/artefactual/atom/blob/qa/2.5.x/plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php#L80-L87
https://github.com/artefactual/atom/blob/qa/2.5.x/plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php#L1228-L1240

You should do the foreach over `$node->creators` and `$node->inheritedCreators`. Only one of then will have data after the constructor. Actually, to avoid having two foreach, you could do it over an array merge of both arrays (and change it in arElasticSearchInformationObjectPdo too, please).

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    if (null !== $partOf = $node->getPartOf())
+    {
+      $data['partOf'] = $partOf;
+    }
```
Otherwise it should be set to null or empty array, like the other data.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    if (array_filter($data))
+    {
+      QubitSearch::getInstance()->partialUpdate(QubitInformationObject::getById($item->id), $data);
+    }
```
We should do the partial update in any case, to remove the fields when needed.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php:
```diff
@@ -1063,6 +1063,50 @@ private function getActRights()
+  {
+    $collectionRootId = $this->getCollectionRootId();
+
+    if ($collectionRootId && $collectionRootId != $this->id)
```
You could reverse the condition and return null directly in here, to avoid having all the code bellow indented to just return null after the if.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObject.class.php:
```diff
@@ -95,6 +111,59 @@ public function recursivelyAddInformationObjects($parentId, $totalRows, $options
+    // If any of the inherited fields have been set to other than null, do partial update
+    if (array_filter($data))
+    {
+      QubitSearch::getInstance()->partialUpdate(QubitInformationObject::getById($item->id), $data);
```
This looks a little overkill in this case, after all the work to improve the population using raw queries, I'd try to avoid having to hydrate the entire IO to do the partial update call. The partial update function only requires the entire resource in the case the document is not found in the index:

https://github.com/artefactual/atom/blob/qa/2.5.x/plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php#L484

I'd change the partialUpdate function, to only require the class of the resource, the id and the partial data. And, if the update fails because the document is not found in the index, I'd do the hydration in there, using the class name and the id before calling the full update (and change the calls from the finding aid job).

jraddaoui replied:
Feel free to disregard this too if running out of time. We can create a new feature ticket to not forgot about the enhancement in the future.

jraddaoui reviewed (changes requested):
Looking better @MikeFE , I've added a few more comments ;)

jraddaoui commented:
Thank you very much for this work @MikeFE!

I'll move to it to a new PR to avoid the new reviewer the need of reading all this comments.

jraddaoui commented:
Moved to #684
